### PR TITLE
[serverless] Support trace context propagation for ALB target groups with MultiValueHeaders

### PR DESCRIPTION
**Pull request by quietsato**
### What does this PR do?

Adds support for extracting trace context from MultiValueHeaders in ALBTargetGroupRequest

### Motivation

Currently, Datadog APM context propagation doesn't work when MultiValueHeaders is enabled in ALB target groups. This PR resolves this issue by allowing trace context propagation even when using MultiValueHeaders configuration.

### Describe how to test/QA your changes

- Enable MultiValueHeaders in ALB target group
- Create a Lambda endpoint
- Send a request with trace context headers to the ALB endpoint
- Verify in Datadog's trace view that the trace context is properly propagated from ALB to Lambda

### Possible Drawbacks / Trade-offs

- Only the first value from MultiValueHeaders is used, subsequent values are ignored

### Additional Notes

- This PR includes test coverage:
  - Unit tests for MultiValueHeadersCarrier
  - Tests for MultiValueHeaders extraction in ALB target groups
  - Tests for MultiValueHeaders support in tag extraction functionality
- Maintains backward compatibility by prioritizing the existing Headers field

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#multi-value-headers-request

### CODE REVIEW

File: pkg/serverless/trace/propagation/carriers.go [hunk 1]
```diff
@@ -253,6 +253,16 @@ func headersCarrier(hdrs map[string]string) (tracer.TextMapReader, error) {
 	return tracer.TextMapCarrier(hdrs), nil
 }
 
+// multiValueHeadersCarrier returns the tracer.TextMapReader used to extract trace
+// context from a MultiValueHeaders field of form map[string][]string.
+func multiValueHeadersCarrier(hdrs map[string][]string) (tracer.TextMapReader, error) {
```
Comment (purple4reina) [inline pkg/serverless/trace/propagation/carriers.go:258]: I suspect you can probably use https://pkg.go.dev/github.com/DataDog/dd-trace-go@v1.24.1/ddtrace/tracer#HTTPHeadersCarrier which won't require you to reinvent the wheel or do extra allocations.

File: pkg/serverless/trigger/extractor.go [hunk 2]
```diff
@@ -188,11 +188,20 @@ func GetTagsFromALBTargetGroupRequest(event events.ALBTargetGroupRequest) map[st
...
-			httpTags["http.referer"] = event.Headers["Referer"]
+
+	headers := event.Headers
+	if headers == nil && event.MultiValueHeaders != nil {
+		headers = make(map[string]string)
+		for k, v := range event.MultiValueHeaders {
+			headers[k] = v[0]
 		}
-		if ua := event.Headers["User-Agent"]; ua != "" {
+	}
```
Comment (purple4reina) [inline pkg/serverless/trigger/extractor.go:198]: You can do this without an extra allocation by checking `event.MultiValueHeaders["Referer"] != nil && len(event.MultiValueHeaders["Referer"]) > 0`

**Review by purple4reina — CHANGES_REQUESTED** [review]
Thanks so much for this PR. It's a great improvement and we really appreciate your willingness to implement it. I've made just a couple of requests for changes but otherwise this looks great.

File: pkg/serverless/trace/propagation/carriers.go (hunk 1 above)
Comment (quietsato) [reply pkg/serverless/trace/propagation/carriers.go]: Thank you for your suggestion!

I've made the changes in the commit 270eb93324937909481c99c7ef943510e129237d to use HTTPHeadersCarrier from dd-trace-go.
I've removed the multiValueHeadersCarrier function and its associated tests, while ensuring that the final context extraction behavior remains covered by the existing tests.
https://github.com/DataDog/datadog-agent/blob/6eda6791cf50470472e9a3616360d2c33be7c2a6/pkg/serverless/trace/propagation/extractor_test.go#L544-L553

File: pkg/serverless/trigger/extractor.go (hunk 2 above)
Comment (quietsato) [reply pkg/serverless/trigger/extractor.go]: Thank you for the suggestion!
I've implemented this improvement in commit aaaa24f04199be385ee9df4553e9e7b4217a236e.

**Comment by purple4reina** [comment]
I reopened this PR as https://github.com/DataDog/datadog-agent/pull/31542 so the tests will run and we can merge it.

**Pull request closed without merging**